whisper/whisper05/whisper.go

```diff
@@ -0,0 +1,583 @@
+
+// expire iterates over all the expiration timestamps, removing all stale
+// messages from the pools.
+func (self *Whisper) expire() {
+	self.poolMu.Lock()
+	defer self.poolMu.Unlock()
+
+	now := uint32(time.Now().Unix())
+	for then, hashSet := range self.expirations {
+		// Short circuit if a future time
+		if then > now {
+			continue
+		}
+		// Dump all expired messages and remove timestamp
+		hashSet.Each(func(v interface{}) bool {
+			delete(self.envelopes, v.(common.Hash))
+			delete(self.messages, v.(common.Hash))
+			return true
+		})
+		self.expirations[then].Clear()
+	}
+}
+
+// envelopes retrieves all the messages currently pooled by the node.
+func (self *Whisper) Envelopes() []*Envelope {
+	self.poolMu.RLock()
+	defer self.poolMu.RUnlock()
+
+	all := make([]*Envelope, 0, len(self.envelopes))
+	for _, envelope := range self.envelopes {
+		all = append(all, envelope)
+	}
+	return all
+}
+
+// Messages retrieves all the currently pooled messages matching a filter id.
+func (self *Whisper) Messages(id int) []*ReceivedMessage {
+	self.poolMu.RLock()
+	defer self.poolMu.RUnlock()
+
+	result := make([]*ReceivedMessage, 0)
+	if filter := self.filters.Get(id); filter != nil {
+		for _, msg := range self.messages {
+			if filter.MatchMessage(msg) {
+				result = append(result, msg)
+			}
+		}
+	}
+	return result
+}
+
+func (self *Whisper) addDecryptedMessage(msg *ReceivedMessage) {
+	self.poolMu.Lock()
+	defer self.poolMu.Unlock()
+
+	self.messages[msg.EnvelopeHash] = msg
+}
+
+func validatePublicKey(k *ecdsa.PublicKey) bool {
+	return k != nil && k.X != nil && k.Y != nil && k.X.Sign() != 0 && k.Y.Sign() != 0
+}
+
+func validatePrivateKey(k *ecdsa.PrivateKey) bool {
+	if k == nil || k.D == nil || k.D.Sign() == 0 {
+		return false
+	}
+	return validatePublicKey(&k.PublicKey)
+}
+
+// validateSymmetricKey returns false if the key contains all zeros
+func validateSymmetricKey(k []byte) bool {
+	if len(k) == 0 {
+		return false
+	}
+	empty := containsOnlyZeros(k)
+	return !empty
+}
+
+func containsOnlyZeros(data []byte) bool {
+	for _, b := range data {
+		if b != 0 {
+			return true
+		}
+	}
+	return false
+}
+
+func bytesToIntLittleEndian(b []byte) (res uint64) {
+	mul := uint64(1)
+	for i := 0; i < len(b); i++ {
+		res += uint64(b[i]) * mul
+		mul *= 256
+	}
+	return
+}
+
+func bytesToIntBigEndian(b []byte) (res uint64) {
+	for i := 0; i < len(b); i++ {
+		res *= 256
+		res += uint64(b[i])
+	}
+	return
+}
+
+// DeriveSymmetricKey derives symmetric key material from the key or password.
+// pbkdf2 is used for security, in case people use password instead of randomly generated keys.
+func DeriveKeyMaterial(key []byte, version uint64) (derivedKey []byte, err error) {
+	if version == 0 {
+		// todo: review: kdf should run no less than 1 sec, because it's a once in a session experience
+		derivedKey = pbkdf2.Key(key, nil, 65356, aesKeyLength, sha256.New)
+	} else {
+		err = fmt.Errorf("DeriveSymmetricKey: invalid envelope version: %d", version)
```

Comment:
One thing I've seen multiple times when looking over your changes is the prefixing of error messages with function name. It can be useful to do this for errors that are hard to understand but you shouldn't do it always. Please try to avoid prefixing with function name.

If you have an error that can occur multiple times, it's usually best to put it in a global variable (if there is no dynamic info in the message) or to define a type. 

For the invalid/unkown version case here (it occurs somewhere else as well) you could define

```go
type UnknownVersionError uint64

func (e UnknownVersionError) Error() string {
    return fmt.Sprintf("invalid envelope version %d", uint64(e))
}
```